File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        adj = [[] for _ in range(numCourses)]
        indegree = [0]*numCourses
        ans =[]

        for pair in prerequisites:
            adj[pair[1]].append(pair[0])
            indegree[pair[0]] +=1
        q =deque()
        for i in range(numCourses):
            if indegree[i]==0:
                q.append(i)
        
        while q:
            curr = q.popleft()
            ans.append(curr)
            for next_course in adj[curr]:
                indegree[next_course] -=1
                if indegree[next_course]==0:
                    q.append(next_course)
        return len(ans)==numCourses

File: 0207-course-schedule/0207-course-schedule.cpp
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses);
        vector<int> indegree(numCourses,0), bfs;

        for(auto &e: prerequisites){
            graph[e[1]].push_back(e[0]);
            indegree[e[0]]++;
    }

        for(int i=0 ;i<numCourses; i++){
            if(indegree[i]==0){
                bfs.push_back(i);
            }
        }

        for(int i=0; i< bfs.size(); i++){
            for(int j : graph[bfs[i]]){
                if(--indegree[j]==0){
                    bfs.push_back(j);
                }
            }
        }
        return bfs.size()==numCourses;
    }
};